### PR TITLE
Unexpected package version warnings

### PR DESCRIPTION
**Pull request by emgarten**
This improves the experience around unexpected package versions and provides helpful warnings around scenarios where the package or project is authored incorrectly without a minimum version on the dependency range, which leads to *floating downwards* and the result is subject to whatever is available on the sources.

I've done some perf testing on this after the changes it looks to be the same with and without the additional checks.

* Fixed a bug in VersionRange.PrettyPrint
* Added AsList extension method to shared extension methods, this is an optimization that was previously under PackageManagement only. 
* FluentAssertions for tests, I've had much better results with using this on top of Xunit, it simplifies comparisions and provides much better messages when failures happen. As part of this I've added new tests using this with names that follow those guidelines. CLI/SDK are using this also.
* Added IRestoreTargetGraph for Mock support
* Minor optimizations/perf improvements around the code changed

Fixes https://github.com/NuGet/Home/issues/5145
Fixes https://github.com/NuGet/Home/issues/5137

### CODE REVIEW

File: src/NuGet.Core/NuGet.Versioning/VersionRangeFormatter.cs [hunk 1]
```diff
@@ -282,68 +281,56 @@ private string GetLegacyString(VersionRangeBase range)
...
+        }
+
+        private void PrettyPrintBound(StringBuilder sb, NuGetVersion version, bool inclusive, string boundChar)
+        {
+            sb.Append(boundChar);
+
+            if (inclusive)
             {
-                sb.Clear();
+                sb.Append("=");
```
Comment (mishra14) [inline src/NuGet.Core/NuGet.Versioning/VersionRangeFormatter.cs:329]: why is this needed?

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/Diagnostics/UnexpectedDependencyMessages.cs [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+{
+    /// <summary>
+    /// Log warnings for packages that did not resolve to the minimum version of the dependency range.
+    /// </summary>
+    public static class UnexpectedDependencyMessages
+    {
+        /// <summary>
+        /// Log warnings for all project issues related to unexpected dependencies.
+        /// </summary>
+        public static async Task Analyze(IEnumerable<IRestoreTargetGraph> graphs, PackageSpec project, ILogger logger)
```
Comment (mishra14) [inline src/NuGet.Core/NuGet.Commands/RestoreCommand/Diagnostics/UnexpectedDependencyMessages.cs]: Rename the method to indicate what is it that is being analyzed.

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/ProjectRestoreCommand.cs [hunk 3]
```diff
@@ -150,31 +152,21 @@ public ProjectRestoreCommand(ProjectRestoreRequest request)
...
 
             // Resolve conflicts
-            _logger.LogVerbose(string.Format(CultureInfo.CurrentCulture, Strings.Log_ResolvingConflicts, name));
+            await _logger.LogAsync(LogLevel.Verbose, string.Format(CultureInfo.CurrentCulture, Strings.Log_ResolvingConflicts, name));
 
             // Flatten and create the RestoreTargetGraph to hold the packages
-            var result = RestoreTargetGraph.Create(runtimeGraph, graphs, context, _logger, framework, runtimeIdentifier);
-
-            // Check if the dependencies got bumped up
-            CheckDependencies(result, _request.Project.Dependencies);
```
Comment (mishra14) [inline src/NuGet.Core/NuGet.Commands/RestoreCommand/ProjectRestoreCommand.cs:169]: what happened to this code? moved?

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/ResolvedDependencyKey.cs [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+
+using System;
+using NuGet.LibraryModel;
+using NuGet.Shared;
+using NuGet.Versioning;
+
+namespace NuGet.Commands
+{
+    /// <summary>
+    /// Parent -(Range) -> Child
```
Comment (mishra14) [inline src/NuGet.Core/NuGet.Commands/RestoreCommand/ResolvedDependencyKey.cs]: Please add some more explanation here.

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/ResolvedDependencyKey.cs [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+        public VersionRange Range { get; }
+
+        /// <summary>
+        /// Child node.
+        /// </summary>
+        public LibraryIdentity Child { get; }
+
+        public ResolvedDependencyKey(LibraryIdentity parent, VersionRange range, LibraryIdentity child)
+        {
+            Parent = parent ?? throw new ArgumentNullException(nameof(parent));
```
Comment (mishra14) [inline src/NuGet.Core/NuGet.Commands/RestoreCommand/ResolvedDependencyKey.cs:37]: 👍 on the `??`

File: src/NuGet.Core/NuGet.Commands/Strings.resx [hunk 6]
```diff
@@ -26,36 +26,36 @@
         <value>[base64 mime encoded string representing a byte array form of the .NET Framework object]</value>
         <comment>This is a comment</comment>
     </data>
-
-    There are any number of "resheader" rows that contain simple
+
```
Comment (mishra14) [inline src/NuGet.Core/NuGet.Commands/Strings.resx:29]: nit: clear stray spaces throughout the comment.

File: src/NuGet.Core/NuGet.Commands/Utility/Extensions.cs [hunk 7]
```diff
@@ -21,15 +23,15 @@ public static ISet<LibraryDependency> GetAllPackageDependencies(this PackageSpec
             // Remove non-package dependencies such as framework assembly references.
             return new HashSet<LibraryDependency>(
                 project.Dependencies.Concat(project.TargetFrameworks.SelectMany(e => e.Dependencies))
-                                    .Where(e => !e.LibraryRange.TypeConstraintAllows(LibraryDependencyTarget.Package)));
```
Comment (mishra14) [inline src/NuGet.Core/NuGet.Commands/Utility/Extensions.cs]: why do we need this inversion?

File: src/NuGet.Core/NuGet.Commands/Utility/Extensions.cs [hunk 8]
```diff
@@ -21,15 +23,15 @@ public static ISet<LibraryDependency> GetAllPackageDependencies(this PackageSpec
...
+                                    .Where(e => e.LibraryRange.TypeConstraintAllows(LibraryDependencyTarget.Package)));
         }
 
         public static ISet<LibraryDependency> GetPackageDependenciesForFramework(this PackageSpec project, NuGetFramework framework)
         {
             // Remove non-package dependencies such as framework assembly references.
             return new HashSet<LibraryDependency>(
                 project.Dependencies.Concat(project.GetTargetFramework(framework).Dependencies)
-                                    .Where(e => !e.LibraryRange.TypeConstraintAllows(LibraryDependencyTarget.Package)));
```
Comment (mishra14) [inline src/NuGet.Core/NuGet.Commands/Utility/Extensions.cs]: same here?

File: src/NuGet.Core/NuGet.Common/Errors/NuGetLogCode.cs [hunk 9]
```diff
@@ -1,13 +1,32 @@
...
+
+        /// <summary>
+        /// Non-exact match on dependency range due to missing package version.
+        /// </summary>
+        NU2503,
+
+        /// <summary>
+        /// Project dependency does not include a lower bound.
+        /// </summary>
+        NU2504,
```
Comment (mishra14) [inline src/NuGet.Core/NuGet.Common/Errors/NuGetLogCode.cs:30]: nit: remove the ending comma

**Review by mishra14 — CHANGES_REQUESTED** [review]
Please address comments and changes where necessary.

File: src/NuGet.Core/NuGet.Versioning/VersionRangeFormatter.cs (hunk 1 above)
Comment (emgarten) [reply src/NuGet.Core/NuGet.Versioning/VersionRangeFormatter.cs:329]: discussed offline

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/Diagnostics/UnexpectedDependencyMessages.cs (hunk 2 above)
Comment (emgarten) [reply src/NuGet.Core/NuGet.Commands/RestoreCommand/Diagnostics/UnexpectedDependencyMessages.cs]: LogMessagesAsync?

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/ProjectRestoreCommand.cs (hunk 3 above)
Comment (emgarten) [reply src/NuGet.Core/NuGet.Commands/RestoreCommand/ProjectRestoreCommand.cs:169]: Moved to Analyze

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/ResolvedDependencyKey.cs (hunk 4 above)
Comment (emgarten) [reply src/NuGet.Core/NuGet.Commands/RestoreCommand/ResolvedDependencyKey.cs]: ``ResolvedDependencyKey represents 'Parent -(Range) -> Child'`` within the graph.

File: src/NuGet.Core/NuGet.Commands/Strings.resx (hunk 6 above)
Comment (emgarten) [reply src/NuGet.Core/NuGet.Commands/Strings.resx:29]: VS2017 does this

File: src/NuGet.Core/NuGet.Commands/Utility/Extensions.cs (hunk 7 above)
Comment (emgarten) [reply src/NuGet.Core/NuGet.Commands/Utility/Extensions.cs]: it was a bug in an earlier commit in this PR

File: test/NuGet.Core.Tests/NuGet.Commands.Test/UnexpectedDependencyMessagesTests.cs [hunk 10]
```diff
@@ -0,0 +1,592 @@
...
+            targetGraph.SetupGet(e => e.Name).Returns("net46/win10");
+            var targetGraphs = new[] { targetGraph.Object };
+            var ignore = new HashSet<string>();
+
+            var logs = UnexpectedDependencyMessages.GetMissingLowerBounds(targetGraphs, ignore).ToList();
+
+            logs.Select(e => e.Message).ShouldBeEquivalentTo(new[]
+            {
+                "x 9.0.0 does not provide an inclusive lower bound for dependency b (<= 5.0.0). An approximate best match of b 2.0.0 was resolved.",
+                "y 8.0.0 does not provide an inclusive lower bound for dependency b (> 1.0.0 && <= 6.0.0). An approximate best match of b 3.0.0 was resolved."
```
Comment (mishra14) [inline test/NuGet.Core.Tests/NuGet.Commands.Test/UnexpectedDependencyMessagesTests.cs:43]: Why is this a warning? 3.0.0 is inside the range (1.0.0, 6.0.0]

Comment (mishra14) [reply test/NuGet.Core.Tests/NuGet.Commands.Test/UnexpectedDependencyMessagesTests.cs:43]: Also, missing codes. I guess those will be plumbed in later?

File: test/NuGet.Core.Tests/NuGet.Commands.Test/UnexpectedDependencyMessagesTests.cs [hunk 11]
```diff
@@ -0,0 +1,592 @@
...
+            var dependency = new ResolvedDependencyKey(parent, VersionRange.Parse("[1.0.0, 3.0.0)"), child);
+            var dependencySet = new HashSet<ResolvedDependencyKey>() { dependency };
+            targetGraph.SetupGet(e => e.ResolvedDependencies).Returns(dependencySet);
+            var targetGraphs = new[] { targetGraph.Object };
+            var ignore = new HashSet<string>();
+
+            await UnexpectedDependencyMessages.Analyze(targetGraphs, project, testLogger);
+
+            testLogger.ShowMessages().Should().NotContain("NU2504");
+            testLogger.ShowMessages().Should().Contain("NU2501");
```
Comment (mishra14) [inline test/NuGet.Core.Tests/NuGet.Commands.Test/UnexpectedDependencyMessagesTests.cs]: Why is this a warning?

File: test/NuGet.Core.Tests/NuGet.Commands.Test/UnexpectedDependencyMessagesTests.cs [hunk 12]
```diff
@@ -0,0 +1,592 @@
...
+            testLogger.ShowMessages().Should().NotContain("NU2502");
+            testLogger.ShowMessages().Should().NotContain("NU2503");
+        }
+
+        [Fact]
+        public async Task GivenAProjectWithMultipleWarningsForXVerifyOnlyOneIsLogged()
```
Comment (mishra14) [inline test/NuGet.Core.Tests/NuGet.Commands.Test/UnexpectedDependencyMessagesTests.cs:113]: How are you asserting that the message is logged only once?

File: test/NuGet.Core.Tests/NuGet.Commands.Test/UnexpectedDependencyMessagesTests.cs [hunk 13]
```diff
@@ -0,0 +1,592 @@
...
+            await UnexpectedDependencyMessages.Analyze(targetGraphs, project, testLogger);
+
+            testLogger.ShowMessages().Should().NotContain("NU2504");
+            testLogger.ShowMessages().Should().NotContain("NU2501");
+            testLogger.ShowMessages().Should().Contain("NU2502");
+            testLogger.ShowMessages().Should().NotContain("NU2503");
+        }
+
+        [Fact]
+        public async Task GivenAProjectWithMultipleWarningsForXVerifyOnlyTheFirstIsLogged()
```
Comment (mishra14) [inline test/NuGet.Core.Tests/NuGet.Commands.Test/UnexpectedDependencyMessagesTests.cs:147]: Again, how are you asserting that only the first ones are logged?

File: test/NuGet.Core.Tests/NuGet.Commands.Test/UnexpectedDependencyMessagesTests.cs [hunk 14]
```diff
@@ -0,0 +1,592 @@
...
+            targetGraph.SetupGet(e => e.Name).Returns("net46/win10");
+            targetGraph.SetupGet(e => e.Framework).Returns(NuGetFramework.Parse("net46"));
+            var targetGraphs = new[] { targetGraph.Object };
+            var ignore = new HashSet<string>();
+
+            UnexpectedDependencyMessages.GetBumpedUpDependencies(targetGraphs, project, ignore).Should().BeEmpty();
+        }
+
+        [Fact]
+        public void GivenAProjectWithABumpedReferenceDependencyVerifyNoMessage()
```
Comment (mishra14) [inline test/NuGet.Core.Tests/NuGet.Commands.Test/UnexpectedDependencyMessagesTests.cs:204]: why is this a non-error case?

File: test/NuGet.Core.Tests/NuGet.Commands.Test/UnexpectedDependencyMessagesTests.cs [hunk 15]
```diff
@@ -0,0 +1,592 @@
...
+            var flattened = new HashSet<GraphItem<RemoteResolveResult>>
+            {
+                new GraphItem<RemoteResolveResult>(new LibraryIdentity("x", NuGetVersion.Parse("2.0.0"), LibraryType.Package))
+            };
+            var targetGraph = new Mock<IRestoreTargetGraph>();
+            targetGraph.SetupGet(e => e.Flattened).Returns(flattened);
+            targetGraph.SetupGet(e => e.Name).Returns("net46/win10");
+            targetGraph.SetupGet(e => e.Framework).Returns(NuGetFramework.Parse("net46"));
+            var targetGraphs = new[] { targetGraph.Object };
+            var ignore = new HashSet<string>() { "X" };
```
Comment (mishra14) [inline test/NuGet.Core.Tests/NuGet.Commands.Test/UnexpectedDependencyMessagesTests.cs:244]: is ignore case insensitive?

File: test/NuGet.Core.Tests/NuGet.Commands.Test/UnexpectedDependencyMessagesTests.cs [hunk 16]
```diff
@@ -0,0 +1,592 @@
...
+            var tfi = GetTFI(NuGetFramework.Parse("net46"), new LibraryRange("x", null, LibraryDependencyTarget.Package));
+            var project = new PackageSpec(tfi)
+            {
+                Name = "proj"
+            };
+
+            var log = UnexpectedDependencyMessages.GetProjectDependenciesMissingLowerBounds(project).Single();
+
+            log.Code.Should().Be(NuGetLogCode.NU2504);
+            log.Message.Should().Be("Project dependency x does not contain an inclusive lower bound. Include a lower bound in the dependency version to ensure consistent restore results.");
```
Comment (mishra14) [inline test/NuGet.Core.Tests/NuGet.Commands.Test/UnexpectedDependencyMessagesTests.cs:382]: I feel that `version range == null` should be a different error than missing lower bound.

**Review by mishra14 — CHANGES_REQUESTED** [review]
review comments

File: test/NuGet.Core.Tests/NuGet.Commands.Test/UnexpectedDependencyMessagesTests.cs (hunk 11 above)
Comment (emgarten) [reply test/NuGet.Core.Tests/NuGet.Commands.Test/UnexpectedDependencyMessagesTests.cs]: discussed offline

File: test/NuGet.Core.Tests/NuGet.Commands.Test/UnexpectedDependencyMessagesTests.cs (hunk 12 above)
Comment (emgarten) [reply test/NuGet.Core.Tests/NuGet.Commands.Test/UnexpectedDependencyMessagesTests.cs:113]: checked by the codes

File: test/NuGet.Core.Tests/NuGet.Commands.Test/UnexpectedDependencyMessagesTests.cs (hunk 13 above)
Comment (emgarten) [reply test/NuGet.Core.Tests/NuGet.Commands.Test/UnexpectedDependencyMessagesTests.cs:147]: codez

File: test/NuGet.Core.Tests/NuGet.Commands.Test/UnexpectedDependencyMessagesTests.cs (hunk 14 above)
Comment (emgarten) [reply test/NuGet.Core.Tests/NuGet.Commands.Test/UnexpectedDependencyMessagesTests.cs:204]: Reference

File: test/NuGet.Core.Tests/NuGet.Commands.Test/UnexpectedDependencyMessagesTests.cs (hunk 15 above)
Comment (emgarten) [reply test/NuGet.Core.Tests/NuGet.Commands.Test/UnexpectedDependencyMessagesTests.cs:244]: yes

File: test/NuGet.Core.Tests/NuGet.Commands.Test/UnexpectedDependencyMessagesTests.cs (hunk 16 above)
Comment (emgarten) [reply test/NuGet.Core.Tests/NuGet.Commands.Test/UnexpectedDependencyMessagesTests.cs:382]: This is more of a test for external callers such as the CLI to ensure we don't break this behavior

**Review by mishra14 — APPROVED** [review]
Discussed offline. Fix the small changes requested then 🚢
